samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,319 @@
+                repos.append(
+                    ECRRepo(logical_id=resource.logical_resource_id, physical_id=resource.physical_resource_id)
+                )
+        return repos
+
+    def get_unreferenced_repos(self) -> List[ECRRepo]:
+        """
+        List deployed ECR repos that is not referenced by current list of functions
+
+        Returns
+        -------
+        List[ECRRepo]
+            List of deployed ECR repos that is not referenced by current list of functions
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        deployed_repos: List[ECRRepo] = self.list_deployed_repos()
+        current_mapping = self._builder.repo_mapping
+
+        unreferenced_repos: List[ECRRepo] = list()
+        for deployed_repo in deployed_repos:
+            for _, current_repo in current_mapping.items():
+                if current_repo.logical_id == deployed_repo.logical_id:
+                    break
+            else:
+                unreferenced_repos.append(deployed_repo)
+        return unreferenced_repos
+
+    def delete_unreferenced_repos(self) -> None:
+        """
+        Blocking call to delete all deployed ECR repos that are unreferenced by a function
+        If repo does not exist, this will simply skip it.
+        """
+        repos = self.get_unreferenced_repos()
+        for repo in repos:
+            try:
+                self._ecr_client.delete_repository(repositoryName=repo.physical_id, force=True)
+            except self._ecr_client.exceptions.RepositoryNotFoundException:
+                LOG.debug("Image repo [%s] not found in companion stack. Skipping deletion.", repo.physical_id)
+
+    def sync_repos(self) -> None:
+        """
+        Blocking call to sync companion stack with the following actions
+        Creates the stack if it does not exist, and updates it if it does.
+        Deletes unreferenced repos if they exist.
+        Deletes companion stack if there isn't any repo left.
+        """
+        has_repo = bool(self.get_repository_mapping())
+        if self.does_companion_stack_exist():
+            self.delete_unreferenced_repos()
+            if has_repo:
+                self.update_companion_stack()
+            else:
+                self.delete_companion_stack()
+        elif has_repo:
+            self.update_companion_stack()
+
+    def does_companion_stack_exist(self) -> bool:
+        """
+        Does companion stack exist
+
+        Returns
+        -------
+        bool
+            Returns True if companion stack exists
+        """
+        try:
+            self._cfn_client.describe_stacks(StackName=self._companion_stack.stack_name)
+            return True
+        except ClientError as e:
+            error_message = e.response.get("Error", {}).get("Message")
+            if error_message == f"Stack with id {self._companion_stack.stack_name} does not exist":
```

Comment:
The error code is ValidationError which can be too generic and might be caused by other issues instead of stack not found.
All the data we have from ClientError is:
```
{
	"response": {
		"Error": {
			"Type": "Sender",
			"Code": "ValidationError",
			"Message": "Stack with id abcdefg-be64f26c-CompanionStack does not exist"
		},
		"ResponseMetadata": {
			"RequestId": "111111-2acf-45a7-98ff-11111111",
			"HTTPStatusCode": 400,
			"HTTPHeaders": {
				"x-amzn-requestid": "111111-2acf-45a7-98ff-11111111",
				"content-type": "text/xml",
				"content-length": "320",
				"date": "Tue, 10 Aug 2021 11:11:11 GMT",
				"connection": "close"
			},
			"RetryAttempts": 0
		}
	},
	"operation_name": "DescribeStacks"
}
```

Reply:
ok, it seems not possible .. thanks